File: services/nuldoc/nuldoc-src/pages/PostPage.ts
import GlobalFooter from "../components/GlobalFooter.ts";
import GlobalHeader from "../components/BlogGlobalHeader.ts";
import PageLayout from "../components/PageLayout.ts";
import TableOfContents from "../components/TableOfContents.ts";
import { Config, getTagLabel } from "../config.ts";
import {
  a,
  article,
  div,
  elem,
  Element,
  h1,
  h2,
  header,
  li,
  ol,
  section,
  ul,
} from "../dom.ts";
import { Document } from "../markdown/document.ts";
import { dateToString } from "../revision.ts";
import { getPostPublishedDate } from "../generators/post.ts";

export default async function PostPage(
  doc: Document,
  config: Config,
): Promise<Element> {
  return await PageLayout({
    metaCopyrightYear: getPostPublishedDate(doc).year,
    metaDescription: doc.description,
    metaKeywords: doc.tags.map((slug) => getTagLabel(config, slug)),
    metaTitle: `${doc.title}｜${config.sites.blog.siteName}`,
    requiresSyntaxHighlight: true,
    site: "blog",
    config,
    children: elem(
      "body",
      { class: "single" },
      GlobalHeader({ config }),
      elem(
        "main",
        { class: "main" },
        article(
          { class: "post-single" },
          header(
            { class: "post-header" },
            h1({ class: "post-title" }, doc.title),
            doc.tags.length !== 0
              ? ul(
                { class: "post-tags" },
                ...doc.tags.map((slug) =>
                  li(
                    { class: "tag" },
                    a(
                      { class: "tag-inner", href: `/tags/${slug}/` },
                      getTagLabel(config, slug),
                    ),
                  )
                ),
              )
              : null,
          ),
          doc.toc && doc.toc.entries.length > 0
            ? TableOfContents({ toc: doc.toc })
            : null,
          div(
            { class: "post-content" },
            section(
              { id: "changelog" },
              h2({}, a({ href: "#changelog" }, "更新履歴")),
              ol(
                {},
                ...doc.revisions.map((rev) =>
                  li(
                    { class: "revision" },
                    elem(
                      "time",
                      { datetime: dateToString(rev.date) },
                      dateToString(rev.date),
                    ),
                    `: ${rev.remark}`,
                  )
                ),
              ),
            ),
            // TODO: refactor
            ...(doc.root.children[0] as Element).children,
          ),
        ),
      ),
      GlobalFooter({ config }),
    ),
  });
}

File: services/nuldoc/lib/nuldoc/pages/post_page.rb
module Nuldoc
  module Pages
    class PostPage
      def initialize(doc:, config:)
        @doc = doc
        @config = config
      end

      def render
        doc = @doc
        config = @config
        Components::PageLayout.new(
          meta_copyright_year: GeneratorUtils.published_date(doc).year,
          meta_description: doc.description,
          meta_keywords: doc.tags.map { |slug| config.tag_label(slug) },
          meta_title: "#{doc.title}｜#{config.sites.blog.site_name}",
          site: 'blog',
          config: config,
          children: DOM::HTMLBuilder.new.build do
            body class: 'single' do
              render Components::BlogGlobalHeader, config: config
              main class: 'main' do
                article class: 'post-single' do
                  header class: 'post-header' do
                    h1(class: 'post-title') { text doc.title }
                    if doc.tags.length.positive?
                      ul class: 'post-tags' do
                        doc.tags.each do |slug|
                          li class: 'tag' do
                            a(class: 'tag-inner', href: "/tags/#{slug}/") { text config.tag_label(slug) }
                          end
                        end
                      end
                    end
                  end
                  render Components::TableOfContents, toc: doc.toc if doc.toc && doc.toc.items.length.positive?
                  div class: 'post-content' do
                    section id: 'changelog' do
                      h2 { a(href: '#changelog') { text '更新履歴' } }
                      ol do
                        doc.revisions.each do |rev|
                          ds = Revision.date_to_string(rev.date)
                          li class: 'revision' do
                            time(datetime: ds) { text ds }
                            text ": #{rev.remark}"
                          end
                        end
                      end
                    end
                    child(*doc.root.children[0].children)
                  end
                end
              end
              render Components::GlobalFooter, config: config
            end
          end
        ).build
      end
    end
  end
end
